syntax = "proto3";
package protos.playbooks;

import "google/protobuf/wrappers.proto";
import "protos/playbooks/source_task_definitions/promql_task.proto";


message Grafana {
  message PromQlMetricExecution {
    message LabelValue {
      google.protobuf.StringValue name = 1;
      google.protobuf.StringValue value = 2;
    }
    google.protobuf.StringValue datasource_uid = 1;
    google.protobuf.StringValue promql_expression = 2;
    repeated LabelValue promql_label_option_values = 3;
    google.protobuf.StringValue dashboard_uid = 4;
    google.protobuf.StringValue dashboard_title = 5;
    google.protobuf.StringValue panel_id = 6;
    google.protobuf.StringValue panel_title = 7;
    google.protobuf.StringValue panel_promql_expression = 8;
    repeated uint32 timeseries_offsets = 9;
  }

  message QueryDashboardPanelMetricTask {
    message Query {
      google.protobuf.StringValue expr = 1;
      google.protobuf.StringValue query = 2;
    } 
    repeated Query queries = 1;
    google.protobuf.StringValue dashboard_id = 2;
    google.protobuf.StringValue panel_id = 3;
    google.protobuf.StringValue datasource_uid = 4;
  }

  message ExecuteAllDashboardPanelsTask {
    google.protobuf.StringValue dashboard_uid = 1;
    google.protobuf.StringValue panel_ids = 2;
    google.protobuf.UInt64Value interval = 3;
    google.protobuf.StringValue template_variables = 4;
  }

  message FetchDashboardVariableLabelValuesTask {
    google.protobuf.StringValue datasource_uid = 1;
    google.protobuf.StringValue label_name = 2;
  }

  message FetchDashboardVariablesTask {
    google.protobuf.StringValue dashboard_uid = 1;
  }

  enum TaskType {
    UNKNOWN = 0;
    PROMQL_METRIC_EXECUTION = 1;
    PROMETHEUS_DATASOURCE_METRIC_EXECUTION = 2;
    QUERY_DASHBOARD_PANEL_METRIC = 3;
    EXECUTE_ALL_DASHBOARD_PANELS = 4;
    FETCH_DASHBOARD_VARIABLE_LABEL_VALUES = 5;
    FETCH_DASHBOARD_VARIABLES = 6;
  }

  TaskType type = 1;
  oneof task {
    PromQlMetricExecution promql_metric_execution = 3;
    PromQl.PromQlMetricExecution prometheus_datasource_metric_execution = 4;
    QueryDashboardPanelMetricTask query_dashboard_panel_metric = 5;
    ExecuteAllDashboardPanelsTask execute_all_dashboard_panels = 6;
    FetchDashboardVariableLabelValuesTask fetch_dashboard_variable_label_values = 7;
    FetchDashboardVariablesTask fetch_dashboard_variables = 8;
  }
}
